src/emmy/tape.cljc

```diff
@@ -317,15 +307,65 @@
    :primal      (.-primal t)
    :in->partial (.-in->partial t)})
 
+(defn tag-of
+  "More permissive version of [[tape-tag]] that returns `nil` when passed a
+  non-perturbation."
+  [x]
+  (cond (tape? x)   (tape-tag x)
+        (d/dual? x) (d/tag x)
+        :else       nil))
+
+(defn inner-tag
+  "Given any number of `tags`, returns the tag most recently bound
+  via [[with-active-tag]] (i.e., the tag connected with the _innermost_ call
+  to [[with-active-tag]]).
+
+  If none of the tags are bound, returns `(apply max tags)`."
+  [& tags]
+  (or (some (apply hash-set tags)
+            d/*active-tags*)
+      (apply max tags)))
+
+(defn tag+perturbation
+  "Given any number of `dxs`, returns a pair of the form
+
+  [<tag> <tape-or-dual-number>]
+
+  containing the tag and instance of [[emmy.differential/Dual]] or [[TapeCell]]
+  associated with the inner-most call to [[with-active-tag]] in the current call
+  stack.
+
+  If none of `dxs` has an active tag, returns `nil`."
+  ([& dxs]
+   (let [m (into {} (mapcat
+                     (fn [dx]
+                       (when-let [t (tag-of dx)]
+                         {t dx})))
```

Comment:
I wonder if it is cheaper to write `[t dx]`. However you choose to write it, it's cool (and I do think that `{t dx}` is a reasonable way to write the lone binding

Reply:
your way is faster!!